Src/DependencyCollector/Shared/ServiceBusDiagnosticsListener.cs

```diff
@@ -0,0 +1,172 @@
+        private const string ExceptionPropertyName = "Exception";
+
+        // We want to reflect in the dependency type that it's one of the queue operations.
+        // This way, UI could have special representation for all kinds of queues
+        // Not knowing particular service
+        private const string DependencyType = "Queue.ServiceBus";
+
+        private readonly TelemetryClient telemetryClient;
+
+        // Every fetcher is unique for event payload and particular property into this payload
+        // when we first receive an event and require particular property, we cache fetcher for it
+        // There are just a few (~10) events we can receive and each will a have a few payload fetchers
+        // I.e. this dictionary is quite small, does not grow up after service warm-up and does not require clean up
+        private readonly ConcurrentDictionary<Property, PropertyFetcher> propertyFetchers = new ConcurrentDictionary<Property, PropertyFetcher>();
+
+        internal ServiceBusDiagnosticsEventHandler(TelemetryConfiguration configuration)
+        {
+            this.telemetryClient = new TelemetryClient(configuration);
+        }
+
+        /// <summary>
+        /// Gets custom handler <see cref="HandleDiagnosticsEvent{TContext}"/>delegate.
+        /// </summary>
+        internal HandleDiagnosticsEvent<HashSet<string>> EventHandler => this.HandleEvent;
+
+        private void HandleEvent(KeyValuePair<string, object> evnt, DiagnosticListener diagnosticListener, HashSet<string> context)
+        {
+            // there is no guarantee IsEnabled was called in case there are multiple listeners in the system. i.e. we need to check it again
+            if (evnt.Key.EndsWith("Start", StringComparison.Ordinal))
+            {
+                return;
+            }
+
+            Activity currentActivity = Activity.Current;
+
+            switch (evnt.Key)
+            {
+                case "Microsoft.Azure.ServiceBus.ProcessSession.Stop":
+                case "Microsoft.Azure.ServiceBus.Process.Stop":
+                    this.OnRequest(evnt.Key, evnt.Value, currentActivity);
+                    break;
+                case "Microsoft.Azure.ServiceBus.Exception":
+                    this.OnException(evnt.Key, evnt.Value);
+                    break;
+                default:
+                    this.OnDependency(evnt.Key, evnt.Value, currentActivity);
+                    break;
+            }
+        }
+
+        private void OnDependency(string name, object payload, Activity activity)
+        {
+            DependencyTelemetry telemetry = new DependencyTelemetry { Type = DependencyType };
+
+            this.SetCommonProperties(name, payload, activity, telemetry);
+
+            // Endpoint is URL of particular ServiceBus, e.g. sb://myservicebus.servicebus.windows.net/
+            telemetry.Data = this.FetchPayloadProperty<Uri>(name, EndpointPropertyName, payload)?.ToString();
```

Comment:
```Endpoint``` has a fixed value fro the whole SB client (is stripped out of connection string) so it does not fir into Data field which should have data specific to particular operation. ```PartitionKey``` would fit better here IMO

Reply:
how related it is to partition?
The data model now is too tight to HTTP, that's why another protocol does not really fit.

I'm fine with moving it to `Endpoint` custom dimension though or removing it all. It contains the servicebus name though, so perhaps custom dimension is good enough

Reply:
Data is a generic purpose field - it can contain URL, SQL query, ID of the data being accessed. So it has data that is vague and not well defined - because of that it will be hard to use it to increase perf of checking endpoint as that will be only limited to a subset of all data (and checking that costs). If we don't see anything useful to put there though let's just leave it empty

I think Endpoint is the most important information about the dependency as it allows to identify where to look next. That's why I believe it should be used as "Target"

Reply:
so if i have 10 queues, what do I do?
Where would you put queue name?